File: Clients/Go/Socket_client/Data/interfaces/Event_State.go
package interfaces

import (
	"encoding/binary"
	"encoding/json"
	"fmt"
	"io"
	"log"
	"math/rand"
	"net"
	"time"
)

type Process struct {
	PID       int              `json:"pid"`
	TTL       int              `json:"ttl"`
	Attempts  int              `json:"attempts"`
	Created   time.Time        `json:"created"`
	Updated   time.Time        `json:"updated"`
	Data      chan interface{} `json:"data"`
	OnTimeout func()           `json:"-"`
}

type ProcessSlice []*Process

var processes ProcessSlice

type ResponseCallback func(response State)

type Event struct {
	Event       string      `json:"event"`
	Destination string      `json:"destination"`
	Data        interface{} `json:"data"`
	Origen      string      `json:"origen"`
	PID         int         `json:"pid"`
}

type State struct {
	Status      bool        `json:"status"`
	Message     string      `json:"message"`
	Error       string      `json:"error"`
	Data        interface{} `json:"data"`
	Destination string      `json:"destination"`
	Origen      string      `json:"origen"`
	PID         int         `json:"pid"`
}

type MessageState struct {
	Status        bool   `json:"status"`
	ServerPID     int    `json:"server_pid"`
	Message       string `json:"state"`
	Error         string `json:"error"`
	ProcessStatus int    `json:"process_status"` // 0=pendiente, 1=en proceso, 2=finalizado
}

func (s State) ToString() string {
	return fmt.Sprintf("State{Status: %v, Message: %q, Error: %q, Data: %v, Destination: %q, Origen: %q, PID: %d}",
		s.Status, s.Message, s.Error, s.Data, s.Destination, s.Origen, s.PID)
}

func (m MessageState) ToString() string {
	return fmt.Sprintf("MessageState{Status: %v, ServerPID: %d, Message: %q, Error: %q}",
		m.Status, m.ServerPID, m.Message, m.Error)
}

func (object Event) SendData(client *Client, timeout *time.Duration, response ...ResponseCallback) error {

	var typeBuf byte = 1
	var pid int = generatePID()
	object.PID = pid

	// Convertir a JSON
	data, err := json.Marshal(object)
	if err != nil {
		return fmt.Errorf("error al convertir el mensaje: %w", err)
	}

	// Tamaño del JSON
	messageSize := uint32(len(data))
	sizeBuffer := make([]byte, 4)
	binary.BigEndian.PutUint32(sizeBuffer, messageSize)

	// Armar el paquete completo: tipo (1B) + tamaño (4B) + datos JSON
	packet := append([]byte{typeBuf}, sizeBuffer...)
	packet = append(packet, data...)

	fmt.Println("intentando Escribir en el servidor")
	// Enviar
	client.WriteMu.Lock()
	_, err = client.Conn.Write(packet)
	client.WriteMu.Unlock()
	if err != nil {
		return fmt.Errorf("error al enviar el mensaje: %w", err)
	}

	fmt.Println("Escrito en el servidor")

	// Registrar proceso para esperar la respuesta
	process := &Process{
		PID:      pid,
		Data:     make(chan interface{}, 1), // bufferizado: no bloquea si ReadData responde rápido
		TTL:      3,
		Attempts: 2,
		Created:  time.Now(),
		Updated:  time.Now(),
	}
	processes = append(processes, process)

	// Esperar respuesta
	if len(response) > 0 {
		if timeout == nil {
			defaultTimeout := 15 * time.Second
			timeout = &defaultTimeout
		}
		select {
		case data := <-process.Data:
			callback := response[0]
			if state, ok := data.(State); ok && state.PID == pid {
				if state.Status {
					callback(state)
					return nil
				} else {
					return fmt.Errorf("error en la respuesta del servidor: %s", state.Error)
				}

			}

			return fmt.Errorf("se esperaba State pero se recibió otro tipo o PID incorrecto")

		case <-time.After(*timeout):
			//Eliminamos el proceso de la lista
			for i, process := range processes {
				if process.PID == pid {
					close(process.Data)                                   // Cerramos el canal para indicar que ya no se enviarán más
					processes = append(processes[:i], processes[i+1:]...) // Eliminamos el proceso de la lista
					log.Println("Proceso eliminado de la lista por que respondio y se encontro:", pid)
					return nil
				}
			}
			return fmt.Errorf("timeout esperando respuesta para PID %d", pid)
		}
	} else {
		go func() {
			select {
			case data := <-process.Data:
				if state, ok := data.(State); ok && state.PID == pid {
					log.Println("Respuesta recibida:", state)
				} else {
					log.Println("No se encontró el proceso o PID incorrecto para la respuesta recibida: ")
					//guardamos en log la respuesta para no perder la información
					log.Println("Respuesta recibida:", data)
				}
			case <-time.After(*timeout):
				log.Printf("Timeout esperando respuesta para PID %d", pid)
			}
		}()
	}
	return nil
}

func (object MessageState) SendData(client *Client) {
	var typeBuf byte = 3

	// Convertir a JSON
	data, err := json.Marshal(object)
	if err != nil {
		log.Println("Error al convertir el mensaje:", err)
		return
	}

	// Tamaño del JSON
	messageSize := uint32(len(data))
	sizeBuffer := make([]byte, 4)
	binary.BigEndian.PutUint32(sizeBuffer, messageSize)

	// Armar el paquete completo: tipo (1B) + tamaño (4B) + datos JSON
	packet := append([]byte{typeBuf}, sizeBuffer...)
	packet = append(packet, data...)

	// Un solo Write
	client.WriteMu.Lock()
	_, err = client.Conn.Write(packet)
	client.WriteMu.Unlock()
	if err != nil {
		log.Println("Error al enviar el mensaje:", err)
	}
	//fmt.Println("MensajeState enviado al servidor:", object.ToString())
}

func (object State) SendData(client *Client, messagePid int, destination string) {

	var stateResponse *MessageState = &MessageState{Message: "El servidor proceso la solicitud", Status: true, ServerPID: messagePid, Error: "", ProcessStatus: 2}
	stateResponse.SendData(client)

	object.Destination = destination

	var typeBuf byte = 2

	// Convertir a JSON
	data, err := json.Marshal(object)
	if err != nil {
		log.Println("Error al convertir el mensaje:", err)
		return
	}

	// Tamaño del JSON
	messageSize := uint32(len(data))
	sizeBuffer := make([]byte, 4)
	binary.BigEndian.PutUint32(sizeBuffer, messageSize)

	// Armar el paquete completo: tipo (1B) + tamaño (4B) + datos JSON
	packet := append([]byte{typeBuf}, sizeBuffer...)
	packet = append(packet, data...)

	// Un solo Write
	client.WriteMu.Lock()
	_, err = client.Conn.Write(packet)
	client.WriteMu.Unlock()
	if err != nil {
		log.Println("Error al enviar el mensaje:", err)
	}
}

func ReadData(conn *Client, clientName string, eventSlice *EventSlice, serverStatus chan bool, latency float64) {
	for {
		// Leer primero tipo + tamaño (9 bytes)
		header := make([]byte, 9)
		_, err := io.ReadFull(conn.Conn, header)
		if err != nil {
			if err == io.EOF {
				log.Println("El servidor cerró la conexión")
				serverStatus <- false
				conn.Close()
				return
			}

			netErr, ok := err.(net.Error)
			if ok && netErr.Temporary() {
				log.Println("Error al leer el encabezado:", err)
				var state *MessageState = &MessageState{Message: "El servidor no puede procesar la solicitud", Status: false, ServerPID: 0, Error: "", ProcessStatus: 2}
				state.SendData(conn)
				continue // 🔴 seguimos con el loop, no cerramos
			}

			// Errores permanentes: notificamos y terminamos
			log.Println("Error al leer encabezado, cerrando conexión:", err)
			serverStatus <- false
			conn.Close()
			return

		} else {

			// Primer byte = tipo
			msgType := header[0]
			// Siguientes 4 bytes = PID
			messagePidBuffer := binary.BigEndian.Uint32(header[1:5])
			//convertimos a int
			messagePid := int(messagePidBuffer)
			// Siguientes 4 bytes = tamaño
			messageSize := binary.BigEndian.Uint32(header[5:9])

			//fmt.Println("Server PID:", messagePid)

			// Leer el JSON completo
			data := make([]byte, messageSize)

			_, err = io.ReadFull(conn.Conn, data)
			if err != nil {
				log.Println("Error al leer el mensaje:", err)
				var state *MessageState = &MessageState{Message: "El servidor no puede procesar la solicitud", Status: false, ServerPID: messagePid, Error: err.Error(), ProcessStatus: 2}
				state.SendData(conn)
			} else {
				switch msgType {
				case 1: // Evento
					var event Event
					if err := json.Unmarshal(data, &event); err != nil {
						log.Println("Error al convertir el JSON:", err)
						break
					}
					//fmt.Println("Evento recibido:", event.Event)
					// lo mandamos al manejador de eventos
					var state *MessageState = &MessageState{Message: "El servidor proceso la solicitud", Status: true, ServerPID: messagePid, Error: "", ProcessStatus: 1}
					state.SendData(conn)
					HandleEvents(event, conn, clientName, eventSlice, int(latency), messagePid)
				case 2: // Estado
					var activeProcess bool = false
					var state State
					if err := json.Unmarshal(data, &state); err != nil {
						log.Println("Error al convertir el JSON:", err)
						break
					}
					var stateResponse *MessageState = &MessageState{Message: "El servidor proceso la solicitud", Status: true, ServerPID: messagePid, Error: "", ProcessStatus: 2}
					stateResponse.SendData(conn)
					found := false
					//fmt.Println("Estado recibido:", state.PID)
					for i, process := range processes {
						if process.PID == state.PID {
							found = true
							process.Data <- state
							close(process.Data)                                   // Cerramos el canal para indicar que ya no se enviarán más
							processes = append(processes[:i], processes[i+1:]...) // Eliminamos el proceso de la lista
							log.Println("Estado enviado al proceso con PID:", process.PID)
							activeProcess = true
							break
						}
					}
					if !found {
						log.Println("No se encontró un proceso con PID:", state.PID)
						log.Println("Respuesta recibida:", state.ToString())
					}
					if !activeProcess {
						// Escribimos en log
						log.Println("No se encontró un proceso activo para el estado:", state.ToString())
					}
				default:
					log.Println("Tipo de mensaje desconocido:", msgType)
				}
			}

		}

	}
}

func generatePID() int {
	rand.Seed(time.Now().UnixNano())
	return rand.Intn(1_000_000)
}

File: Clients/Python/Socket_client/Data/interfaces/Event_State.py
"""Core messaging structures and transport helpers for the Python client."""

from __future__ import annotations

import json
import logging
import queue
import random
import socket
import struct
import threading
import time
from dataclasses import dataclass, field
from typing import TYPE_CHECKING, Any, Callable, Dict, Optional


if TYPE_CHECKING:  # pragma: no cover
    from .EventSlice import EventSlice
    from Client import Client

logger = logging.getLogger(__name__)

ResponseCallback = Callable[["State"], None]

StatusQueue = queue.Queue
StatusQueueEmpty = queue.Empty


@dataclass
class Process:
    pid: int
    ttl: int = 3
    attempts: int = 2
    created: float = field(default_factory=time.time)
    updated: float = field(default_factory=time.time)
    data: "queue.Queue[object]" = field(default_factory=lambda: queue.Queue(maxsize=1))
    on_timeout: Optional[Callable[[], None]] = None

    def deliver(self, payload: object) -> None:
        self.updated = time.time()
        try:
            self.data.put_nowait(payload)
        except queue.Full:  # pragma: no cover - buffer is size 1 by design
            pass

    def await_data(self, timeout: float) -> object:
        self.updated = time.time()
        return self.data.get(timeout=timeout)


_processes: Dict[int, Process] = {}
_process_lock = threading.RLock()


def _register_process(process: Process) -> None:
    with _process_lock:
        _processes[process.pid] = process


def _remove_process(pid: int) -> Optional[Process]:
    with _process_lock:
        return _processes.pop(pid, None)


def _deliver_process(pid: int, payload: object) -> bool:
    with _process_lock:
        process = _processes.get(pid)
    if process is None:
        return False
    process.deliver(payload)
    _remove_process(pid)
    return True


def _read_exact(conn: "Client", length: int) -> bytes:
    buffer = bytearray()
    while len(buffer) < length:
        chunk = conn._conn.recv(length - len(buffer))
        if not chunk:
            raise EOFError("connection closed while reading")
        buffer.extend(chunk)
    return bytes(buffer)


def generate_pid() -> int:
    return random.randint(0, 999_999)


@dataclass
class Event:
    event: str
    destination: str
    data: Any
    origen: str = ""
    pid: int = 0

    def to_dict(self) -> Dict[str, Any]:
        return {
            "event": self.event,
            "destination": self.destination,
            "data": self.data,
            "origen": self.origen,
            "pid": self.pid,
        }

    @classmethod
    def from_dict(cls, payload: Dict[str, Any]) -> "Event":
        return cls(
            event=payload.get("event", ""),
            destination=payload.get("destination", ""),
            data=payload.get("data"),
            origen=payload.get("origen", ""),
            pid=int(payload.get("pid", 0)),
        )

    def send_data(
        self,
        client: Client,
        timeout: Optional[float] = None,
        callback: Optional[ResponseCallback] = None,
    ) -> Optional["State"]:
        self.pid = generate_pid()

        payload = json.dumps(self.to_dict()).encode("utf-8")
        message_size = len(payload)
        packet = struct.pack(">BI", 1, message_size) + payload
        process = Process(pid=self.pid)
        _register_process(process)

        try:
            with client._write_lock:
                client._conn.sendall(packet)
        except OSError as exc:
            _remove_process(self.pid)
            raise RuntimeError("error al enviar el mensaje") from exc

        effective_timeout = 15.0 if timeout is None else timeout
        print("Event sent, waiting for response...")  # Debug print
        if callback is None:
            threading.Thread(
                target=_fire_and_log,
                args=(process, self.pid, effective_timeout),
                daemon=True,
            ).start()
            return None

        try:
            data = process.await_data(effective_timeout)
        except queue.Empty as exc:
            _remove_process(self.pid)
            raise TimeoutError(f"timeout esperando respuesta para PID {self.pid}") from exc

        state = _ensure_state(data, self.pid)
        if state.status:
            callback(state)
            return state

        raise RuntimeError(f"error en la respuesta del servidor: {state.error}")


@dataclass
class State:
    data: Any
    pid: int=0
    status: bool = True
    message: str = ""
    error: str = ""
    destination: str=""
    origen: str=""
    

    def to_dict(self) -> Dict[str, Any]:
        return {
            "status": self.status,
            "message": self.message,
            "error": self.error,
            "data": self.data,
            "destination": self.destination,
            "origen": self.origen,
            "pid": self.pid,
        }

    @classmethod
    def from_dict(cls, payload: Dict[str, Any]) -> "State":
        return cls(
            status=bool(payload.get("status", False)),
            message=str(payload.get("state", "")),
            error=str(payload.get("error", "")),
            data=payload.get("data"),
            destination=str(payload.get("destination", "")),
            origen=str(payload.get("origen", "")),
            pid=int(payload.get("pid", 0)),
        )

    def __str__(self) -> str:
        return (
            "State{Status: %s, Message: %r, Error: %r, Data: %r, Destination: %r, Origen: %r, PID: %d}"
            % (self.status, self.message, self.error, self.data, self.destination, self.origen, self.pid)
        )

    def send_data(self, client: Client, message_pid: int,dstination: str) -> None:
        ack = MessageState(
            status=True,
            server_pid=message_pid,
            message="El servidor proceso la solicitud",
            error="",
            process_status=1,
        )
        ack.send_data(client)

        self.destination = dstination
        self.pid = message_pid

        payload = json.dumps(self.to_dict()).encode("utf-8")
        message_size = len(payload)
        packet = struct.pack(">BI", 2, message_size) + payload
        try:
            with client._write_lock:
                client._conn.sendall(packet)
        except OSError as exc:
            logger.error("Error al enviar el mensaje: %s", exc)


@dataclass
class MessageState:
    status: bool
    server_pid: int
    message: str
    error: str
    process_status: int

    def to_dict(self) -> Dict[str, Any]:
        return {
            "status": self.status,
            "server_pid": self.server_pid,
            "state": self.message,
            "error": self.error,
            "process_status": self.process_status,
        }

    def __str__(self) -> str:
        return (
            "MessageState{Status: %s, ServerPID: %d, Message: %r, Error: %r}"
            % (self.status, self.server_pid, self.message, self.error)
        )

    def send_data(self, client: "Client") -> None:
        payload = json.dumps(self.to_dict()).encode("utf-8")
        message_size = len(payload)
        packet = struct.pack(">BI", 3, message_size) + payload
        try:
            with client._write_lock:
                client._conn.sendall(packet)
        except OSError as exc:
            logger.error("Error al enviar el mensaje: %s", exc)


def _ensure_state(data: object, pid: int) -> State:
    if isinstance(data, State):
        if data.pid != pid:
            raise RuntimeError("PID incorrecto en la respuesta")
        return data
    if isinstance(data, dict):
        return State.from_dict(data)
    raise RuntimeError("Respuesta inesperada recibida")


def _fire_and_log(process: Process, pid: int, timeout: float) -> None:
    try:
        data = process.await_data(timeout)
        state = _ensure_state(data, pid)
        logger.info("Respuesta recibida: %s", state)
    except queue.Empty:
        logger.warning("Timeout esperando respuesta para PID %d", pid)
    except Exception as exc:  # noqa: BLE001
        logger.warning("Respuesta inesperada para PID %d: %s", pid, exc)
    finally:
        _remove_process(pid)


def read_data(
    conn: "Client",
    client_name: str,
    event_slice: "EventSlice",
    server_status: "StatusQueue[bool]",
    latency: float,
) -> None:
    while True:
        try:
            header = _read_exact(conn, 9)
        except EOFError:
            logger.info("El servidor cerró la conexión")
            _notify_status(server_status, False)
            conn._conn.close()
            return
        except OSError as exc:
            logger.error("Error al leer encabezado, cerrando conexión: %s", exc)
            _notify_status(server_status, False)
            conn._conn.close()
            return

        msg_type = header[0]
        message_pid = struct.unpack(">I", header[1:5])[0]
        message_size = struct.unpack(">I", header[5:9])[0]

        try:
            payload = _read_exact(conn, message_size)
        except (EOFError, OSError) as exc:
            logger.error("Error al leer el mensaje: %s", exc)
            failure = MessageState(
                status=False,
                server_pid=message_pid,
                message="El servidor no puede procesar la solicitud",
                error=str(exc),
                process_status=2,
            )
            failure.send_data(conn)
            continue

        if msg_type == 1:  # Evento
            try:
                event_payload = json.loads(payload.decode("utf-8"))
                event = Event.from_dict(event_payload)
            except json.JSONDecodeError as exc:
                logger.error("Error al convertir el JSON: %s", exc)
                continue

            ack = MessageState(
                status=True,
                server_pid=message_pid,
                message="El servidor proceso la solicitud",
                error="",
                process_status=1,
            )
            ack.send_data(conn)

            try:
                from .Handlers import handle_events  # Local import to avoid circular dependency

                handle_events(event, conn, client_name, event_slice, int(latency), message_pid)
            except Exception as exc:  # noqa: BLE001
                logger.exception("Error al procesar el evento %s: %s", event.event, exc)
        elif msg_type == 2:  # Estado
            try:
                state_payload = json.loads(payload.decode("utf-8"))
                state = State.from_dict(state_payload)
            except json.JSONDecodeError as exc:
                logger.error("Error al convertir el JSON: %s", exc)
                continue

            ack = MessageState(
                status=True,
                server_pid=message_pid,
                message="El servidor proceso la solicitud",
                error="",
                process_status=2,
            )
            ack.send_data(conn)

            if _deliver_process(state.pid, state):
                logger.info("Estado enviado al proceso con PID: %d", state.pid)
            else:
                logger.warning("No se encontró un proceso con PID: %d", state.pid)
                logger.info("Respuesta recibida: %s", state)
        else:
            logger.warning("Tipo de mensaje desconocido: %d", msg_type)


def _notify_status(queue_: "StatusQueue[bool]", status: bool) -> None:
    try:
        queue_.put_nowait(status)
    except queue.Full:  # pragma: no cover - queue size is unbounded by default
        pass


__all__ = [
    "Event",
    "State",
    "MessageState",
    "Process",
    "ResponseCallback",
    "read_data",
    "generate_pid",
    "StatusQueue",
    "StatusQueueEmpty",
]
